### PR TITLE
[SD-1439] Invalid-SQL-statement-when-filtering-Products-via-Storefront-API

### PR DESCRIPTION
**Pull request by kgorazd**
https://spark-solutions.atlassian.net/browse/SD-1439

### CODE REVIEW

File: core/app/models/spree/variant.rb [hunk 1]
```diff
@@ -60,7 +60,7 @@ class Variant < Spree::Base
 
     after_touch :clear_in_stock_cache
 
-    scope :in_stock, -> { joins(:stock_items).where('count_on_hand > ? OR track_inventory = ?', 0, false) }
+    scope :in_stock, -> { joins(:stock_items).where('spree_stock_items.count_on_hand > ? OR track_inventory = ?', 0, false) }
```
Comment (damianlegawiec) [inline core/app/models/spree/variant.rb:63]:
```suggestion
    scope :in_stock, -> { joins(:stock_items).where("#{Spree::StockItem.table_name}.count_on_hand > ? OR track_inventory = ?", 0, false) }
```

**Review by damianlegawiec — COMMENTED** [review]
@kgorazd also we need a regression test that we'll test the (previously) failing scenario

**Review by pmilewski — APPROVED** [review]

File: core/app/models/spree/variant.rb [hunk 2]
```diff
@@ -60,7 +60,7 @@ class Variant < Spree::Base
 
     after_touch :clear_in_stock_cache
 
-    scope :in_stock, -> { joins(:stock_items).where('count_on_hand > ? OR track_inventory = ?', 0, false) }
+    scope :in_stock, -> { joins(:stock_items).where("#{Spree::StockItem.table_name}.count_on_hand > ? OR track_inventory = ?", 0, false) }
```
Comment (damianlegawiec) [inline core/app/models/spree/variant.rb:63]: @kgorazd we should also add the table name for track_inventory attribute as well

File: core/spec/models/spree/product/scopes_spec.rb [hunk 3]
```diff
@@ -318,4 +318,11 @@
       expect(products_by_store).to contain_exactly(product)
     end
   end
+
+  # Regression test for SD-1439 ambiguous column name: count_on_hand
+  describe '#in_stock.in_stock_or_backorderable' do
```
Comment (damianlegawiec) [inline core/spec/models/spree/product/scopes_spec.rb:323]: @kgorazd besides unit test we need a request specs with the same parameters as passed in the sentry error that we caught - https://github.com/spree/spree/blob/main/api/spec/requests/spree/api/v2/storefront/products_spec.rb

**Review by damianlegawiec — APPROVED** [review]

**Comment by damianlegawiec** [comment]
@kgorazd this still fails on Ruby 3 - https://app.circleci.com/pipelines/github/spree/spree/4379/workflows/ed5f1ab1-1270-4c99-ba49-e6a8081c1943/jobs/39474/tests#failed-test-0

**Comment by kgorazd** [comment]
@damianlegawiec 
Adding a request specs with the same parameters as passed in the sentry error caused this error on MySQL.
`ActiveRecord::StatementInvalid:
  Mysql2::Error: Duplicate column name 'id'`
It was related to query produced by sorting by price and SKU at once. `SELECT DISTINCT spree_products.*, spree_prices.amount, spree_products.*, spree_variants.sku`.

**Review by damianlegawiec — APPROVED** [review]

**Pull request merged**